src/main/scala/viper/gobra/ast/frontend/Ast.scala

```diff
@@ -416,12 +423,33 @@ case class PExpCompositeVal(exp: PExpression) extends PCompositeVal // exp is ne
 
 case class PLitCompositeVal(lit: PLiteralValue) extends PCompositeVal
 
-case class PFunctionLit(args: Vector[PParameter], result: PResult, body: PBlock) extends PLiteral with PCodeRootWithResult with PScope
+case class PFunctionLit(decl: PClosureNamedDecl) extends PLiteral with PActualExpression
+
+case class PClosureNamedDecl(id: Option[PIdnDef], decl: PClosureDecl) extends PGhostMisc {
+  val result: PResult = decl.result
+}
+
+case class PClosureDecl(args: Vector[PParameter],
+                  result: PResult,
+                  spec: PFunctionSpec,
+                  body: Option[(PBodyParameterInfo, PBlock)]) extends PFunctionOrClosureDecl with PCodeRootWithResult with PActualMisc
+
+case class PClosureSpecInstance(func: PNameOrDot, params: Vector[PClosureSpecParameter]) extends PGhostMisc
+
+case class PClosureSpecParameter(key: Option[PClosureSpecParameterKey], exp: PExpression) extends PGhostMisc
+
+case class PClosureSpecParameterKey(name: String) extends PGhostMisc
+
+case class PClosureImplements(closure: PExpression, spec: PClosureSpecInstance) extends PGhostExpression
+
+case class PClosureImplProof(impl: PClosureImplements, block: PBlock) extends PGhostStatement with PScope
 
 case class PInvoke(base: PExpressionOrType, args: Vector[PExpression]) extends PActualExpression
 
 // TODO: Check Arguments in language specification, also allows preceding type
 
+case class PCallWithSpec(base: PExpression, args: Vector[PExpression], spec: PClosureSpecInstance) extends PActualExpression
```

Comment:
I would maybe add a spec option to invoke to reduce redundancy. However, I am not very sure about this.